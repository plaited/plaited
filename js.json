{
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "plaited",
    "paths": {
      "plaited/jsx-runtime": [
        "./src/jsx/runtime.ts"
      ],
      "plaited/jsx-dev-runtime": [
        "./src/jsx/dev-runtime.ts"
      ]
    }
  }
}