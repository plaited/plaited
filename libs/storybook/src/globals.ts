import { global } from '@storybook/global';

const { window: globalWindow } = global;


declare global {
  interface Window {
    STORYBOOK_ENV: 'plaited';
  }
}

if (globalWindow) {
  globalWindow.STORYBOOK_ENV = 'plaited';
}
