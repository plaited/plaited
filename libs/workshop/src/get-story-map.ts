import path from 'node:path'
import { trueTypeOf, kebabCase } from '@plaited/utils'
import { PlaitedElement } from 'plaited'
import { Story, StoryMap } from './types.js'
import { toId } from './utils.js'

export const getStoryMap = async (
  entryPoints: string[],
  srcDir: string
) => {
  const storyMap: StoryMap = new Map()
  const titles = new Set<string>()
  for(const entry of entryPoints) {
    const { default: meta, ...stories } = await import(entry)
    // P1 handle bad meta export
    if(
      trueTypeOf(meta) !== 'object' 
    ) {
      console.error(
        `Export: [ default ] in ${entry} \n Default export is not a story meta config`
      )
      continue
    }
    if(
      !Object.hasOwn(meta, 'title') ||
      !Object.hasOwn(meta, 'description') ||
      !Object.hasOwn(meta,'template')
    ) {
      console.error(
        `Export: [ default ] in ${entry} \n Default export is missing key values pairs`
      )
      continue
    }
    const title = meta.title
    // P2 handle incorrectly formatted titles
    if (!/^[a-zA-Z][a-zA-Z/0-9]*$/.test(title)) {
      console.error(
        `Invalid title "${title}" in ${entry} \n Title must only include alphanumeric characters delineated by a "/"`
      )
      continue
    }
    const normalizedTitle = title.toLowerCase()
    // P3 handle duplicate normalized titles
    if(titles.has(normalizedTitle)) {
      console.error(
        `Rename meta: [ ${title} ] in ${entry} \n Title already in use`
      )
      continue
    }
    titles.add(normalizedTitle)
    for (const name in stories) {
      // P4 handle incorrectly formatted names
      if (!/^[a-zA-Z][a-zA-Z_0-9]*$/.test(name)) {
        console.error(
          `Invalid name "${name}" in ${entry} \n Name must only include alphanumeric characters delineated by a "_"`
        )
        continue
      }
      // P5 handle non story exports
      const story = stories[name] as Story
      if(
        trueTypeOf(story) !== 'object'
      ) {
        console.error(
          `Exported Story: [ ${name} ] in ${entry} \n Export [ ${name} ] is not a story`
        )
        continue
      }
      if(
        !('attrs' in story) ||
        !('description' in story)
      ) {
        console.error(
          `Exported Story: [ ${name} ] in ${entry} \n Export [ ${name} ] is missing key values pairs`
        )
        continue
      }
      const id = toId(title, name)
      // P6 handle duplicate normalized names
      if(storyMap.has(id)) {
        console.error(
          `Rename story: [ ${name} ] in ${entry} \n KebabCase  [${kebabCase(name)}] already in use`
        )
        continue
      }
      storyMap.set(id, {
        ...story,
        play: Boolean(story.play),
        name,
        clientPath: `/${path.relative(srcDir, entry).replace(/\.tsx?$/, '.js')}`,
        template: meta.template as PlaitedElement,
        srcPath: entry,
        title: meta.title as string,
      })
    }
  }
  return  storyMap
}
