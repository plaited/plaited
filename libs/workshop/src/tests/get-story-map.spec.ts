import { test, expect, beforeEach, afterEach  } from'@jest/globals'
import path from 'node:path'
import fg from 'fast-glob'
import sinon from 'sinon'
import { getStoryMap } from '../get-story-map.js'


let logStub
let logMessages

beforeEach(() => {
  // Initialize logMessages array
  logMessages = []

  // Create a stub on console.log
  logStub = sinon.stub(console, 'error')

  // Replace console.log implementation to store log messages
  logStub.callsFake((...args) => {
    logMessages.push(args.join(' '))
  })
})

afterEach(() => {
  // Restore console.log to its original implementation
  logStub.restore()
})
test('getStoryMap: valid', async () => {
  const srcDir = `${process.cwd()}/src/tests/__mocks__/get-story-valid`
  const entryPoints = await fg(path.resolve(srcDir, '**/*.stories.{tsx,ts}'))

  const storyMap = await getStoryMap(entryPoints, srcDir)
  expect(storyMap).toMatchSnapshot()
})

test('getStoryMap: bad-meta', async () => {
  const srcDir = `${process.cwd()}/src/tests/__mocks__/get-story-invalid/bad-meta`
  const entryPoints = await fg(path.resolve(srcDir, '**/*.stories.{tsx,ts}'))
  const storyMap = await getStoryMap(entryPoints, srcDir)
  expect(logMessages.length).toBe(2)
  expect(logMessages.some(el => el.endsWith('is missing key values pairs'))).toBeTruthy()
  expect(logMessages.some(el => el.startsWith('Export: [ default ] '))).toBeTruthy()
  expect(storyMap.size).toBe(0)
})

test('getStoryMap: bad-title', async () => {
  const srcDir = `${process.cwd()}/src/tests/__mocks__/get-story-invalid/bad-title`
  const entryPoints = await fg(path.resolve(srcDir, '**/*.stories.{tsx,ts}'))
  const storyMap = await getStoryMap(entryPoints, srcDir)
  expect(logMessages.length).toBe(1)
  expect(logMessages[0].startsWith('Invalid title')).toBeTruthy()
  expect(storyMap.size).toBe(0)
})

test('getStoryMap: duplicate-title', async () => {
  const srcDir = `${process.cwd()}/src/tests/__mocks__/get-story-invalid/duplicate-title`
  const entryPoints = await fg(path.resolve(srcDir, '**/*.stories.{tsx,ts}'))
  const storyMap = await getStoryMap(entryPoints, srcDir)
  expect(logMessages.length).toBe(1)
  expect(logMessages[0].startsWith('Rename meta')).toBeTruthy()
  expect(storyMap.size).toBe(2)
})

test('getStoryMap: bad-story', async () => {
  const srcDir = `${process.cwd()}/src/tests/__mocks__/get-story-invalid/bad-story`
  const entryPoints = await fg(path.resolve(srcDir, '**/*.stories.{tsx,ts}'))
  const storyMap = await getStoryMap(entryPoints, srcDir)
  expect(logMessages.length).toBe(3)
  expect(logMessages[0].startsWith('Invalid name')).toBeTruthy()
  expect(logMessages[1].startsWith('Exported Story: [ ')).toBeTruthy()
  expect(logMessages[2].endsWith('missing key values pairs')).toBeTruthy()
  expect(storyMap.size).toBe(0)
})

test('getStoryMap: duplicate-stories', async () => {
  const srcDir = `${process.cwd()}/src/tests/__mocks__/get-story-invalid`
  const entryPoints = await fg(path.resolve(srcDir, 'duplicate-stories.stories.tsx'))
  const storyMap = await getStoryMap(entryPoints, srcDir)
  expect(logMessages.length).toBe(1)
  expect(logMessages[0].startsWith('Rename story:')).toBeTruthy()
  expect(storyMap.size).toBe(1)
})
