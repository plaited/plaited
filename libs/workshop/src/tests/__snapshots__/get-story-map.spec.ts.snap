// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getStoryMap: valid 1`] = `
Map {
  "example-stories--play-story" => {
    "attrs": {
      "children": "Example story",
    },
    "clientPath": "/example.stories.js",
    "description": "replaces placeholder text",
    "name": "playStory",
    "play": true,
    "srcPath": "/home/edward/Workspace/plaited/libs/workshop/src/tests/__mocks__/get-story-valid/example.stories.tsx",
    "template": [Function],
    "title": "Example/Stories",
  },
  "example-stories--regular-story" => {
    "attrs": {},
    "clientPath": "/example.stories.js",
    "description": "renders with placeholder",
    "name": "regularStory",
    "play": false,
    "srcPath": "/home/edward/Workspace/plaited/libs/workshop/src/tests/__mocks__/get-story-valid/example.stories.tsx",
    "template": [Function],
    "title": "Example/Stories",
  },
  "example-nested-stories--play-story" => {
    "attrs": {
      "children": {
        "content": "<input type="text" value="default value"/>",
        "stylesheets":         Set {},
      },
    },
    "clientPath": "/nested/example.stories.js",
    "description": "nested mock description",
    "name": "playStory",
    "play": true,
    "srcPath": "/home/edward/Workspace/plaited/libs/workshop/src/tests/__mocks__/get-story-valid/nested/example.stories.tsx",
    "template": [Function],
    "title": "Example/NestedStories",
  },
  "example-nested-stories--regular-story" => {
    "attrs": {
      "children": "nope",
    },
    "clientPath": "/nested/example.stories.js",
    "description": "nested mock description",
    "name": "regularStory",
    "play": false,
    "srcPath": "/home/edward/Workspace/plaited/libs/workshop/src/tests/__mocks__/get-story-valid/nested/example.stories.tsx",
    "template": [Function],
    "title": "Example/NestedStories",
  },
}
`;
